EPIPE error on Windows runs

## Bug

- Cypress v3.0.1
- Windows
- Node 10.1.0
- npm 6.1.0

Saw this error on AppVeyor Windows CI https://ci.appveyor.com/project/cypress-io/cypress-example-todomvc/build/1.0.169

```
events.js:167
      throw er; // Unhandled 'error' event
      ^
Error: shutdown EPIPE
    at ReadStream.Socket._final (net.js:369:25)
    at callFinal (_stream_writable.js:615:10)
    at process._tickCallback (internal/process/next_tick.js:63:19)
Emitted 'error' event at:
    at ReadStream.onerror (_stream_readable.js:693:12)
    at ReadStream.emit (events.js:182:13)
    at emitErrorNT (internal/streams/destroy.js:82:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:50:3)
    at process._tickCallback (internal/process/next_tick.js:63:19)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @ cypress:run: `cypress run`
```

Maybe

1. a long request that’s still in flight and being proxied when the server shuts down (since the tests pass)
2. shutting down the http-server when there are open connections
3. killing the child processes (?)

Need to debug with `DEBUG=cypress:*` to isolate the problem

With debug output the last messages are

```
cypress:server:run final results of all runs: ...
cypress:server:cypress about to exit with code 0 +93ms
events.js:167
      throw er; // Unhandled 'error' event
      ^
Error: shutdown EPIPE
    at ReadStream.Socket._final (net.js:369:25)
    at callFinal (_stream_writable.js:615:10)
    at process._tickCallback (internal/process/next_tick.js:63:19)
Emitted 'error' event at:
    at ReadStream.onerror (_stream_readable.js:693:12)
    at ReadStream.emit (events.js:182:13)
    at emitErrorNT (internal/streams/destroy.js:82:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:50:3)
    at process._tickCallback (internal/process/next_tick.js:63:19)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @ cypress:run: `cypress run`
```

Hmm, found an interesting thing - there is an error from bundling step @chrisbreiding 
```
  cypress:server:controllers:spec send C:\Users\appveyor\AppData\Roaming\Cypress\cy\production\projects\cypress-example-todomvc-31c529e7c99c2a4beb3947eb6590b09f\bundles\cypress\integration\app_spec.js +1s
  cypress:browserify errored bundling C:\Users\appveyor\AppData\Roaming\Cypress\cy\production\projects\cypress-example-todomvc-31c529e7c99c2a4beb3947eb6590b09f\bundles\cypress\integration\integration\app_spec.js { Error: Cannot find module 'C:\projects\cypress-example-todomvc\cypress\integration\integration\app_spec.js' from 'C:\projects\cypress-example-todomvc'
    at C:\Users\appveyor\AppData\Local\Cypress\Cache\3.0.1\Cypress\resources\app\packages\server\node_modules\browser-resolve\node_modules\resolve\lib\async.js:55:21
    at load (C:\Users\appveyor\AppData\Local\Cypress\Cache\3.0.1\Cypress\resources\app\packages\server\node_modules\browser-resolve\node_modules\resolve\lib\async.js:69:43)
    at onex (C:\Users\appveyor\AppData\Local\Cypress\Cache\3.0.1\Cypress\resources\app\packages\server\node_modules\browser-resolve\node_modules\resolve\lib\async.js:92:31)
    at C:\Users\appveyor\AppData\Local\Cypress\Cache\3.0.1\Cypress\resources\app\packages\server\node_modules\browser-resolve\node_modules\resolve\lib\async.js:22:47
    at C:\Users\appveyor\AppData\Local\Cypress\Cache\3.0.1\Cypress\resources\app\packages\server\node_modules\graceful-fs\polyfills.js:284:29
    at FSReqWrap.oncomplete (fs.js:152:21)
  stream: 
   Labeled {
     _readableState: 
      ReadableState {
        objectMode: true,
        highWaterMark: 16,
        buffer: BufferList { length: 0 },
        length: 0,
        pipes: [Object],
        pipesCount: 1,
        flowing: true,
        ended: false,
        endEmitted: false,
        reading: true,
        sync: false,
        needReadable: true,
        emittedReadable: false,
        readableListening: false,
        resumeScheduled: false,
        destroyed: false,
        defaultEncoding: 'utf8',
        awaitDrain: 0,
        readingMore: false,
        decoder: null,
        encoding: null },
     readable: true,
     domain: null,
     _events: 
      { end: [Array],
        error: [Function],
        data: [Function: ondata],
        _mutate: [Object] },
     _eventsCount: 4,
     _maxListeners: undefined,
     _writableState: 
      WritableState {
        objectMode: true,
        highWaterMark: 16,
        finalCalled: false,
        needDrain: false,
        ending: true,
        ended: true,
        finished: true,
        destroyed: false,
        decodeStrings: true,
        defaultEncoding: 'utf8',
        length: 0,
        writing: false,
        corked: 0,
        sync: false,
        bufferProcessing: false,
        onwrite: [Function],
        writecb: null,
        writelen: 0,
        bufferedRequest: null,
        lastBufferedRequest: null,
        pendingcb: 0,
        prefinished: true,
        errorEmitted: false,
        bufferedRequestCount: 0,
        corkedRequestsFree: [Object] },
     writable: false,
     allowHalfOpen: true,
     _options: { objectMode: true },
     _wrapOptions: { objectMode: true },
     _streams: [ [Object] ],
     length: 1,
     label: 'deps' },
  filePath: 'C:\\projects\\cypress-example-todomvc\\cypress\\integration\\integration\\app_spec.js',
  originalStack: 'Error: Cannot find module \'C:\\projects\\cypress-example-todomvc\\cypress\\integration\\integration\\app_spec.js\' from \'C:\\projects\\cypress-example-todomvc\'\n    at C:\\Users\\appveyor\\AppData\\Local\\Cypress\\Cache\\3.0.1\\Cypress\\resources\\app\\packages\\server\\node_modules\\browser-resolve\\node_modules\\resolve\\lib\\async.js:55:21\n    at load (C:\\Users\\appveyor\\AppData\\Local\\Cypress\\Cache\\3.0.1\\Cypress\\resources\\app\\packages\\server\\node_modules\\browser-resolve\\node_modules\\resolve\\lib\\async.js:69:43)\n    at onex (C:\\Users\\appveyor\\AppData\\Local\\Cypress\\Cache\\3.0.1\\Cypress\\resources\\app\\packages\\server\\node_modules\\browser-resolve\\node_modules\\resolve\\lib\\async.js:92:31)\n    at C:\\Users\\appveyor\\AppData\\Local\\Cypress\\Cache\\3.0.1\\Cypress\\resources\\app\\packages\\server\\node_modules\\browser-resolve\\node_modules\\resolve\\lib\\async.js:22:47\n    at C:\\Users\\appveyor\\AppData\\Local\\Cypress\\Cache\\3.0.1\\Cypress\\resources\\app\\packages\\server\\node_modules\\graceful-fs\\polyfills.js:284:29\n    at FSReqWrap.oncomplete (fs.js:152:21)\n' } +15ms
  cypress:browserify finished bundling C:\Users\appveyor\AppData\Roaming\Cypress\cy\production\projects\cypress-example-todomvc-31c529e7c99c2a4beb3947eb6590b09f\bundles\cypress\support\index.js +63ms
```

Same error was also in other failed build runs, just lost in all other noise.

Got the same error on Windows on node version 8! Running tests with `cypress open` works fine.
```
events.js:183
      throw er; // Unhandled 'error' event
      ^

Error: shutdown EPIPE
    at _errnoException (util.js:1022:11)
    at Socket.onSocketFinish (net.js:307:25)
    at emitNone (events.js:111:20)
    at Socket.emit (events.js:208:7)
    at finishMaybe (_stream_writable.js:614:14)
    at endWritable (_stream_writable.js:622:3)
    at Socket.Writable.end (_stream_writable.js:573:5)
    at Socket.end (net.js:509:31)
    at Socket.onend (_stream_readable.js:595:10)
    at Object.onceWrapper (events.js:313:30)
    at emitNone (events.js:111:20)
    at Socket.emit (events.js:208:7)
    at endReadableNT (_stream_readable.js:1064:12)
    at _combinedTickCallback (internal/process/next_tick.js:138:11)
    at process._tickCallback (internal/process/next_tick.js:180:9)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! wecapangular@0.0.0 cypress:run: `cypress run`
npm ERR! Exit status 1
```

```
Angular CLI: 6.0.5
Node: 8.11.1
OS: win32 x64
Angular: 6.0.3
```

Im seeing the exact same error as @kayvanbree 
trying to realse on vsts agent
npm/5.6.0 node/v8.11.2 win32 x64

2018-06-08T10:08:23.1708695Z ##[error]events.js:183
      throw er; // Unhandled 'error' event
      ^
Error: shutdown EPIPE
    at _errnoException (util.js:992:11)
    at Socket.onSocketFinish (net.js:307:25)
    at emitNone (events.js:111:20)
    at Socket.emit (events.js:208:7)
    at finishMaybe (_stream_writable.js:614:14)
    at endWritable (_stream_writable.js:622:3)
    at Socket.Writable.end (_stream_writable.js:573:5)
    at Socket.end (net.js:499:31)
    at Socket.onend (_stream_readable.js:595:10)
    at Object.onceWrapper (events.js:313:30)
    at emitNone (events.js:111:20)
    at Socket.emit (events.js:208:7)
    at endReadableNT (_stream_readable.js:1064:12)
    at _combinedTickCallback (internal/process/next_tick.js:138:11)
    at process._tickCallback (internal/process/next_tick.js:180:9)

cypress:server:cypress about to exit with code 0 +17s
events.js:167
      throw er; // Unhandled 'error' event
      ^

Error: shutdown EPIPE
    at ReadStream.Socket._final (net.js:366:25)
    at callFinal (_stream_writable.js:612:10)
    at process._tickCallback (internal/process/next_tick.js:63:19)
Emitted 'error' event at:
    at ReadStream.onerror (_stream_readable.js:687:12)
    at ReadStream.emit (events.js:182:13)
    at emitErrorNT (internal/streams/destroy.js:82:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:50:3)
    at process._tickCallback (internal/process/next_tick.js:63:19)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! ffe@1.0.0 cypress:run: `cypress run`
npm ERR! Exit status 1

--- a/cli/lib/exec/spawn.js
+++ b/cli/lib/exec/spawn.js
@@ -108,6 +108,18 @@ module.exports = {
           process.stderr.write(data)
         })
 
+        // https://github.com/cypress-io/cypress/issues/1841
+        // In some versions of node, it will throw on windows
+        // when you close the parent process after piping
+        // into the child process
+        process.stdin.on('error', (err) => {
+          if (err.code === 'EPIPE') {
+            return
+          }
+
+          throw err
+        })
+
         if (options.detached) {
           child.unref()
         }

--- a/cli/lib/exec/spawn.js
+++ b/cli/lib/exec/spawn.js
@@ -111,7 +111,9 @@ module.exports = {
         // https://github.com/cypress-io/cypress/issues/1841
         // In some versions of node, it will throw on windows
         // when you close the parent process after piping
-        // into the child process
+        // into the child process. unpiping does not seem
+        // to have any effect. so we're just catching the
+        // error here and not doing anything.
         process.stdin.on('error', (err) => {
           if (err.code === 'EPIPE') {
             return

--- a/cli/test/lib/exec/spawn_spec.js
+++ b/cli/test/lib/exec/spawn_spec.js
@@ -313,5 +313,35 @@ describe('lib/exec/spawn', function () {
         expect(process.stderr.write).not.to.be.calledWith(buf1)
       })
     })
+
+    it('catches process.stdin errors and returns when code=EPIPE', function () {
+      this.spawnedProcess.on.withArgs('close').yieldsAsync(0)
+
+      return spawn.start()
+      .then(() => {
+        const fn = () => {
+          const err = new Error()
+          err.code = 'EPIPE'
+          return process.stdin.emit('error', err)
+        }
+
+        expect(fn).not.to.throw
+      })
+    })
+
+    it('throws process.stdin errors code!=EPIPE', function () {
+      this.spawnedProcess.on.withArgs('close').yieldsAsync(0)
+
+      return spawn.start()
+      .then(() => {
+        const fn = () => {
+          const err = new Error('wattttt')
+          err.code = 'FAILWHALE'
+          return process.stdin.emit('error', err)
+        }
+
+        expect(fn).to.throw(/wattttt/)
+      })
+    })
   })
 })

--- a/cli/test/lib/exec/spawn_spec.js
+++ b/cli/test/lib/exec/spawn_spec.js
@@ -4,6 +4,7 @@ const cp = require('child_process')
 const os = require('os')
 const tty = require('tty')
 const path = require('path')
+const EE = require('events')
 
 const state = require(`${lib}/tasks/state`)
 const xvfb = require(`${lib}/exec/xvfb`)
@@ -35,6 +36,7 @@ describe('lib/exec/spawn', function () {
         on: sinon.stub().returns(undefined),
       },
     }
+    sinon.stub(process, 'stdin').value(new EE)
     sinon.stub(cp, 'spawn').returns(this.spawnedProcess)
     sinon.stub(xvfb, 'start').resolves()
     sinon.stub(xvfb, 'stop').resolves()
@@ -319,13 +321,17 @@ describe('lib/exec/spawn', function () {
 
       return spawn.start()
       .then(() => {
+        let called = false
+
         const fn = () => {
+          called = true
           const err = new Error()
           err.code = 'EPIPE'
           return process.stdin.emit('error', err)
         }
 
-        expect(fn).not.to.throw
+        expect(fn).not.to.throw()
+        expect(called).to.be.true
       })
     })
 

